---
title: "Angsd Project Final"
author: "Aditi Gopalan"
date: "04/26/2022"
output: html_document
code_folding: hide
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
```{css, echo = FALSE}
/* Styles for YAML title, author and header  */
h1.title {
  font-size: 35px;
  color: DarkBlue;
  text-align: center;

}
h4.author { 
    font-size: 18px;
    text-align: center;
    color: DarkBlue;
}
h4.date { 
  font-size: 15px;
  text-align: center;
  color: DarkBlue;
}
/* Styles for other report elements  */
h1 {
  font-size: 30px;
  color: DarkBlue;
}
h2 {
  font-size: 20px;
  color: DarkBlue;
}
/* Table captions (header) */
caption {
      color: Black;
      font-size: 1.3em;
}
/* Figure captions (legend) */
p.caption {
  font-size: 1.3em;
}
```
# Introduction 

The increase of lipid generation, storage and uptake to support rapid growth is common in many cancers. Studies have shown that high cholesterol and a high fat diet accelerate cancer development. Lipid metabolism pathways are tweaked in cancers in order to promote tumor growth and metastasis. 

Pyruvate kinase (PK) is the enzyme that catalyzes the last step of glycolysis by converting phosphoenolpyruvate (PEP) to pyruvate. PKM2 gene codes for the PK enzyme. In this study, PKM2 knockout caused an increase in plasma cholesterol levels, which in turn causes tumor progression.

TMEM33 is an ER transmembrane protein. Induction of TMEM33 in liver resulted in a significant increase of total plasma cholesterol levels.

The data I analyzed was selected from the paper PKM2-TMEM33 axis regulates lipid homeostasis in cancer cells. I hypothesized that knocking out of PKM would produce significantly different results in the promotion of cancer through and increase in lipid storage and uptake than any of the other genotypes. 

I was initially interested in seeing if expression of human endogenous transposable elements were linked with cancer progression, but eventually found this to be a more riveting direction to steer my project.


# Results 

## PCA 
From the results, we actually do see a clear clustering of different genotypes. As hypothesized, we see that PKM2 knockout causes a huge difference. Since PKM2 codes for PK enzyme which is an essential part of the glyolysis pathway, it makes sense that there is a clear distinction. 

![Figure 1: PCA Plot](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/PCA.jpg)

## Heatmaps
I then plotted a heatmap of sample to sample distances to compare similiarities between samples. Again, we see that that PKM2 KO clusters significantly differently from the rest. 

![Figure 2: Sample-Sample Heatmap](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/sample_to_sample.jpg)

I was curious to check the top 30 differentially expressed genes and if they vary within samples and if that can be correlated to cancer progression as well. Here are a few genes that I picked off the heatmap that were differentially expression in PKM KO and were linked to promoting carcinogensis. 

![Figure 3: Gene to Sample Heatmap](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/Gene_to_sample.jpg)

![Table 1: Gene Functions](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/table.jpg)

## Volcano Plot

Next, I checked the differential expression of the genes using a volcano plot for the condition PKM2KO vs PKM2 WT and noticed that 56000+ genes were differentially expressed. 

Some significant differences are FOXA1 which is known to be linked with EMT and cancer metastasis, PIK3CD which is linked with migration, invasion and growth of CRCs and GRHL1 which is linked to promotion of cell proliferation.

![Figure 4: Volcano Plot PKM WT vs PKM KO](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/DEG_Volcano.jpg)

## Enrichment Analyses

Finally, I did GO Enrichment analysis, just to see which pathways were enriched. It make sense that pathways strongly linked to energetics were suppressed like "cellular response to lippolysaccharide". While this was sort of helpful, I thought KEGG enrichment may be more insightful 

![Figure 5: GO Enrichment Analysis](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/GO_Enrichment.jpg)

![Figure 5: KEGG Enrichment Analysis](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/KEGG_Enriched .jpg)

It makes sense that pathways related to lipid metabolism were enriched like Alcoholism. It is also interesting to see Notch Signalling enrichment and TNF Signalling suppression, since these pathways are strongly linked to cancer, might be interesting to see if lipid metabolism effects these in any ways moving forward.

# Methods 

## Method Overview 

1. Download 

2. FastQC + MultiQC

3. STAR

4. featureCounts

5. DESeq2

6. Analysis and Interpretation

## 1. Downloading

The SRA study number linked to the paper of my interest is SRP174612. https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=SRP322205&o=acc_s%3Aa

I downloaded the accession file for the study and saved it as "runs.txt"

These were the SRA numbers for the runs, renamed the files as follows 
```
SRR14704380 > NT1
SRR14704381
SRR14704382
SRR14704383 > KO1 
SRR14704384
SRR14704385
SRR14704386 > NT1_1
SRR14704387
SRR14704388
SRR14704389 > KO1_1
SRR14704390
SRR14704391
SRR14704392 >VI_1
SRR14704393
SRR14704394
SRR14704395> VI1_1
SRR14704396
SRR14704397
SRR14704398
SRR14704399
SRR14704400
```
I then downloaded the files using the following bash script (download.sh)
```
#! /bin/bash -l

# Batch commands

#SBATCH --partition=angsd_class
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --job-name=down_ad
#SBATCH --time=48:00:00
#SBATCH --mem=60G
#SBATCH --mail-user=adg4001@med.corell.edu
#SBATCH --mail-type=END,FAIL

echo "Starting at:" `date` >> log.txt
echo "This is job #:" $SLURM_JOB_ID >> log.txt
echo "Running on node:" `hostname` >> log.txt
echo "Running on cluster:" $SLURM_CLUSTER_NAME >> log.txt
echo "This job was assigned the temporary (local) directory:" $TMPDIR >> log.txt

spack load sra-toolkit@2.10.7

for i in $(cat $'runs.txt'); do prefetch ${i}; fastq-dump --gzip --skip-technical --readids --read-filter pass --dumpbase --clip ${i}; done
```

## 2. QC

Running fastqc and MultiQC on the files using the bash script "fastq.sh"

```
#! /bin/bash -l

#SBATCH --partition=angsd_class
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --job-name=fastq_ad
#SBATCH --time=48:00:00
#SBATCH --mem=60G
#SBATCH --mail-user=adg4001@med.corell.edu
#SBATCH --mail-type=END,FAIL

echo "Starting at:" `date` >> log.txt
echo "This is job #:" $SLURM_JOB_ID >> log.txt
echo "Running on node:" `hostname` >> log.txt
echo "Running on cluster:" $SLURM_CLUSTER_NAME >> log.txt
echo "This job was assigned the temporary (local) directory:" $TMPDIR >> log.txt
spack load fastqc
spack load -r py-multiqc

find . -name "*.gz" | xargs -n 1 fastqc --extract

multiqc .
```

The multiqc results look fine! 

![1](C:/Users/aditi/Desktop/angsd/Angsd Project/fastqc_per_base_sequence_quality_plot.png)

![2](C:/Users/aditi/Desktop/angsd/Angsd Project/fastqc_per_base_n_content_plot.png)
![3](C:/Users/aditi/Desktop/angsd/Angsd Project/fastqc_per_sequence_gc_content_plot.png)
![4](C:/Users/aditi/Desktop/angsd/Angsd Project/fastqc_per_sequence_quality_scores_plot.png)
![5](C:/Users/aditi/Desktop/angsd/Angsd Project/fastqc_sequence_duplication_levels_plot.png)

## 3. STAR Alignment 

I then aligned the reads using "star.sh"

```
#! /bin/bash -l

#SBATCH --partition=angsd_class
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --job-name=star_ad
#SBATCH --time=48:00:00
#SBATCH --mem=60G
#SBATCH --mail-user=adg4001@med.corell.edu
#SBATCH --mail-type=END,FAIL

echo "Starting at:" `date` >> log.txt
echo "This is job #:" $SLURM_JOB_ID >> log.txt
echo "Running on node:" `hostname` >> log.txt
echo "Running on cluster:" $SLURM_CLUSTER_NAME >> log.txt
echo "This job was assigned the temporary (local) directory:" $TMPDIR >> log.txt

spack load star@2.7.0e
spack load samtools@1.8%gcc@6.3.0

# STAR: Map reads against the indexed genome
for i in $(cat $'runs.txt');
do STAR --runMode alignReads \
        --genomeDir /athena/angsd/scratch/adg4001/refIndex \
        --readFilesIn ${i}_pass.fastq.gz \
        --readFilesCommand zcat \
        --outFileNamePrefix /athena/angsd/scratch/adg4001/star/${i}. \
        --outSAMtype BAM SortedByCoordinate;
done

for i in $(ls *.bam); do samtools index ${i}; done
```

## 4. Running FeatureCounts

I ran featurcounts using the bash file "counts.sh" attached below
ye
```
#SBATCH --partition=angsd_class
#SBATCH --nodes=1
#SBATCH --ntasks=16
#SBATCH --job-name=adg4001_counts
#SBATCH --time=48:00:00
#SBATCH --mem=60G
#SBATCH --mail-user=adg4001@med.cornell.edu
#SBATCH --mail-type=END,FAIL

echo "Starting at:" `date` >> log.txt
echo "This is job #:" $SLURM_JOB_ID >> log.txt
echo "Running on node:" `hostname` >> log.txt
echo "Running on cluster:" $SLURM_CLUSTER_NAME >> log.txt
echo "This job was assigned the temporary (local) directory:" $TMPDIR >> log.txt
spack load subread

featureCounts -f -O -p \
              -t exon \
                -a /athena/angsd/scratch/adg4001/refIndex/refGenome/gencode.v39.annotation.gtf  \
              -o featurecounts/counts.txt \
              /athena/angsd/scratch/adg4001/star/*.bam \
```
I scped the results to my local

## Reading the counts


```{r}

library(ggplot2)
theme_set(theme_bw(base_size = 16)) # for making plots
library(magrittr) # for "pipe"-like coding in R (R < v.4.1)
library(reshape)
library(dplyr)

folder <- "C:/Users/aditi/Desktop/angsd/ANGSD"

counts <- read.delim('counts.txt.summary', sep='\t')

names(counts) <- gsub(".Aligned.sortedByCoord.out.bam", "", names(counts))
names(counts) <- gsub("alignments.", "", names(counts))
names(counts) <- gsub("X.athena.angsd.scratch.adg4001.star.", "", names(counts))

df <- function(d){
  d <- d[d$Status %in% c('Assigned', 'Unassigned_NoFeatures', 'Unassigned_Ambiguity'),]
  df <- data.frame()
  
  for(i in 2:dim(d)[2]){
    for(j in 1:length(d[,i])){
      df <- rbind(df, c(colnames(d)[i], d$Status[j], d[j, i]))
    }
  }
  names(df) <- c('Sample', 'Status', '# Reads')
  df$`# Reads` <- as.numeric(df$`# Reads`)
  return(df)
}
counts_finals <- df(counts)

#Plotting
library(ggplot2)
ggplot(counts_finals, aes(fill=Status, x=`Sample`, y=`# Reads`)) + 
           geom_bar(position="dodge", stat="identity") +
           ggtitle('Feature Counts') +
           coord_flip()
```
The assigned reads for all conditions are significantly greater 

## 5. Running DESeq2

```{r, eval = FALSE}
library(DESeq2)

counts <- read.table('C:/Users/aditi/Desktop/angsd/ANGSD/counts.txt', header=T)
counts_df <- as.data.frame(counts)
genes <- counts_df$Geneid


library(EnsDb.Hsapiens.v79)

length(genes)
genes <- gsub("\\..*","",genes)

counts_df$Geneid <- genes
counts_df <- counts_df[!duplicated(counts_df$Geneid),]
dim(counts_df)
geneIDs1 <- ensembldb::select(EnsDb.Hsapiens.v79, keys= genes, keytype = "GENEID", columns = "SYMBOL")

f_l <- geneIDs1$SYMBOL
drop_list <- geneIDs1$GENEID
counts2 <- counts_df[counts_df$Geneid %in% drop_list, ]

counts <- counts[,7:ncol(counts)]
#:unts <- counts[-1:-2:-3]

names(counts) <- gsub(x=names(counts), pattern=".Aligned.sortedByCoord.out.bam", replacement="")
names(counts) <- gsub(x=names(counts), pattern="alignments.", replacement="")
names(counts) <- gsub(x=names(counts), pattern="X.athena.angsd.scratch.adg4001.star.", replacement="")

names(counts2) <- gsub(x=names(counts2), pattern=".Aligned.sortedByCoord.out.bam", replacement="")
names(counts2) <- gsub(x=names(counts2), pattern="alignments.", replacement="")
names(counts2) <- gsub(x=names(counts2), pattern="X.athena.angsd.scratch.adg4001.star.", replacement="")

colData <- read.csv('C:/Users/aditi/Desktop/angsd/ANGSD/SraRunTable.txt', sep=',')
colData <- colData[, c('Run', 'Genotype', 'source_name')]
colData$Run <- factor(colData$Run)
colData$Genotype <- factor(colData$Genotype)
colData$source_name <- factor(colData$source_name)

counts2 <- counts2[-1:-5]
counts2 <- counts2[-1]


dat <- DESeqDataSetFromMatrix(countData = counts2,
                              colData = colData,
                              design = ~ Genotype)
dds <- DESeq(dat)

```
## 6. Analysis and Interpretation

### MA Plot

![MA Plot](C:/Users/aditi/Desktop/angsd/ANGSD/Plots/MAPlot.jpg)


### PCA
```{r, eval = FALSE}
vsd <- vst(dds, blind=FALSE)
plotPCA(vsd, intgroup=c('Genotype'))
```
### Heatmaps
```{r, eval = FALSE}
sampleDists <- dist(t(assay(vsd)))

library("RColorBrewer")
library("pheatmap")
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$Genotype, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)

ntd <- normTransform(dds)
rld <- rlog(dds)

expmatrix_DESeq <- DESeq2::rlog(dds, fitType="local")
expmatrix <- SummarizedExperiment::assay(expmatrix_DESeq)

select <- order(rowMeans(expmatrix),decreasing=TRUE)[1:30]
df <- data.frame(Genotype = SummarizedExperiment::colData(dds)[,c("Genotype")], row.names = rownames(SummarizedExperiment::colData(dds)))
pheatmap::pheatmap(expmatrix[select,], cluster_rows=FALSE, show_rownames=TRUE, cluster_cols=TRUE, annotation_col=df)
```
### Volcano 

```{r, eval = FALSE}
library(EnhancedVolcano)
res <- results(dds)

head(res[order(res$padj),], 10)

#res <- results(dds, name="Genotype TMEM33.WT vs PKM2.WT")
#res <- results(dds, ,
  #x = 'log2FoldChange',contrast=c("Genotype","TMEM33.WT","PKM2.WT"))

resultsNames(dds)
head(rowData(dds),10)
resLFC <- lfcShrink(dds, coef="Genotype_PKM2KO_vs_PKM2.WT", type="apeglm")

plotMA(resLFC, ylim=c(-2,2))

rownames(resLFC) <- f_l

EnhancedVolcano(resLFC,
    lab = rownames(resLFC),
    x = 'log2FoldChange',
    y = 'pvalue')

               
```
### Enrichment Analysis
```{r, eval = FALSE}
library(clusterProfiler)
library(enrichplot)
# we use ggplot2 to add x axis labels (ex: ridgeplot)
library(ggplot2)

organism = "org.Hs.eg.db"
library(organism, character.only = TRUE)

# we want the log2 fold change 
original_gene_list <- resLFC$log2FoldChange

# name the vector
names(original_gene_list) <- f_l

# omit any NA values 
gene_list<-na.omit(original_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
gene_list = sort(gene_list, decreasing = TRUE)

gse <- gseGO(geneList=gene_list, 
             ont ="ALL", 
             keyType = "SYMBOL", 
             nPerm = 10000, 
             minGSSize = 3, 
             maxGSSize = 800, 
             pvalueCutoff = 0.05, 
             verbose = TRUE, 
             OrgDb = organism, 
             pAdjustMethod = "none")
require(DOSE)
dotplot(gse, showCategory=10, split=".sign") + facet_grid(.~.sign)

```

```{r, eval = FALSE}
# Convert gene IDs for gseKEGG function
# We will lose some genes here because not all IDs will be converted
ids<-bitr(names(original_gene_list), fromType = "ENSEMBL", toType = "ENTREZID", OrgDb=organism)
 # remove duplicate IDS (here I use "ENSEMBL", but it should be whatever was selected as keyType)
dedup_ids = ids[!duplicated(ids[c("ENSEMBL")]),]

# Create a new dataframe df2 which has only the genes which were successfully mapped using the bitr function above
df2 = resLFC[drop_list %in% dedup_ids$ENSEMBL,]

# Create a new column in df2 with the corresponding ENTREZ IDs
df2$Y = dedup_ids$ENTREZID

# Create a vector of the gene unuiverse
kegg_gene_list <- df2$log2FoldChange

# Name vector with ENTREZ ids
names(kegg_gene_list) <- df2$Y

# omit any NA values 
kegg_gene_list<-na.omit(kegg_gene_list)

# sort the list in decreasing order (required for clusterProfiler)
kegg_gene_list = sort(kegg_gene_list, decreasing = TRUE)

kegg_organism = "hsa"
kk2 <- gseKEGG(geneList     = kegg_gene_list,
               organism     = kegg_organism,
               nPerm        = 10000,
               minGSSize    = 3,
               maxGSSize    = 800,
               pvalueCutoff = 0.05,
               pAdjustMethod = "none",
               keyType       = "ncbi-geneid")

dotplot(kk2, showCategory = 10, title = "Enriched Pathways" , split=".sign") + facet_grid(.~.sign)

```

# Discussion 

Knocking out PKM2 definitely indicated increase in carcinogensis. However, I didn’t see any strong differences with respect to TMEM33, even though it is a downstream effector of PKM2. More insights into the exact mechanisms of cancer and signalling in these pathways could provide more insights. 

Since the Notch Signalling pathway and TNF-Signalling pathway are strongly linked to cancer, I think it may be interesting to dive into how these pathways have any ties to cancer progression in the context of energetics. 

# References 

1. Liu, Fabao, et al. "PKM2‐TMEM33 axis regulates lipid homeostasis in cancer cells by controlling SCAP stability." The EMBO Journal 40.22 (2021): e108065.
2. Hanahan, Douglas, and Robert A. Weinberg. "Hallmarks of cancer: the next generation." cell 144.5 (2011): 646-674.
3. https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#wald-test-individual-steps
4. https://bioconductor.org/packages/devel/bioc/vignettes/EnhancedVolcano/inst/doc/EnhancedVolcano.html
5. https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html