---
title: "Quant Homework 4"
author: "Aditi Gopalan"
date: "20/03/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Problem 1 (Easy)

Provide a symbolic formula that defines a causal polymorphism AND provide a definition of a causal polymorphism in words. Use no more than two total sentences for your answer.

```
A position in the genome where an experimental manipulation of the DNA would produce an effect on the phenotype under specifiable condition is a casual mutation. Causal polymorphism is when this condition results in the existence of one or more alleles at a locus.
```
Consider a case where we have rejected the null hypothesis for a genotype in a GWAS analysis. What are reasons why this genotype may NOT indicate the position of a causal polymorphism (provide two reasons total)?

```
When we reject the null hypothesis, we assume that we have located a position in the genome that contains a causal polymorphism. The reasons it may NOT indicate the position of a causal polymorphism may be due to linkage disequilibrium i.e.,
1. the genotypes not physically linked 2. The genotypes are not close to each other
```
## 2

Problem 2 (Medium)
Write code that inputs the phenotype, plus an (additional) line of code that calculates the number of samples n and report the number n (NOTE: you do not have to output the phenotypes (!!) just provide the R code and report the value for n.

```
NOTE: I used code from lab 7 and 8 for this hw
```

```{r P2}
phenotypes <- read.table("QG22 - hw4_phenotypes.txt", col.names=c("Height"))
N_samples = nrow(phenotypes)
paste0("Number of samples: ", N_samples)
```

Write code that produces a histogram of your phenotype data (NOTE: provide the R code and the histogram).

```{r p2.1}
hist(phenotypes$Height, main="Histogram of Heights", xlab="Height")
```
Write code that inputs the genotype data plus a line of code that outputs the number of genotypes N and sample size n and report these numbers (NOTE: that you do not have to output the genotypes (!!) just provide the R code and report the value for N and n you obtained from these data).

```{r }
genotypes <- read.table("QG22 - hw4_genotypes.txt")
N_gen = ncol(genotypes)
paste0("Number of genotypes: ", N_gen)
paste0("Sample size: ", nrow(genotypes)/2)
```
Write code that converts your genotype data input in part [c] into two new matrices, the first a matrix where each genotype is converted to the appropriate Xa value and the second where each genotype is converted to the appropriate Xd value (NOTE: that you do not have to output the matrices (!!) just provide the R code that will create the matrices if we run it).

```{r}
genotype_coder <- function(geno_import){

  xa_converter <- function(geno_col, numSamples, maf_limit){
    geno_count <- table(geno_col)
    minor_allele <- names(geno_count[geno_count == min(geno_count)])
    #must now undo the stacking together of the mapply and match up the positions for each person
    xa <- (geno_col[1:numSamples]==minor_allele) + (geno_col[(numSamples+1):length(geno_col)]==minor_allele)
    xa <- xa-1
    return(xa)
  }
  xa_mat  <- apply(geno_import, 2, xa_converter, nrow(geno_import)/2, 0.05)
  xd_mat <- 1 - 2*abs(xa_mat)
  
  return(list(xa_mat,xd_mat))
}

res <- genotype_coder(genotypes)
xa_mat <- res[[1]]
xd_mat <- res[[2]]

```

Write code to calculate MLE(β^)=[β^μ,β^a,β^d] for each genotype in the dataset, an F-statistic for each genotype, and a p-value for each genotype using the R function pf(F-statistic, df1, df2, lower.tail = FALSE). PLEASE NOTE (!!): that you may NOT use an existing R function for ANY of these calculations other than the calculation of the p-value (=you must write code that calculates each component except the p-value) and NOTE: you do not have to output anything (!!) just provide the R code.

```{r}
library(MASS)

pval_calculator <- function(pheno_input, xa_input, xd_input){ 
  
  n_samples <- length(xa_input)
  X_mx <- cbind(1,xa_input,xd_input)
  MLE_beta <- ginv(t(X_mx) %*% X_mx) %*% t(X_mx) %*% pheno_input 
  y_hat <- X_mx %*% MLE_beta
  SSM <- sum((y_hat - mean(pheno_input))^2) 
  SSE <- sum((pheno_input - y_hat)^2)
  df_M <- 2
  df_E <- n_samples - 3
  MSM <- SSM / df_M 
  MSE <- SSE / df_E
  Fstatistic <- MSM / MSE
      # to check if it is correct
  pval <- pf(Fstatistic, df_M, df_E,lower.tail = FALSE)
  return(pval) 
}

pval_mx <- rep(0,ncol(xa_mat)) 

for(i in 1:ncol(xa_mat)){
  pval_mx[i] <- pval_calculator(phenotypes[,1], xa_mat[,i], xd_mat[,i]) 
}
```
Write code to produce a Manhattan plot (i.e., genotypes in order on the x-axis and -log(p- values) on the y-axis. PLEASE NOTE (!!): do NOT use an R function (=write your own code to produce the Manhattan plot) but DO provide your code AND your Manhattan plot.

```{r}
library(ggplot2)

plot_df <- data.frame(index = 1:length(pval_mx), pval = pval_mx) 
ggplot(plot_df, aes(index, -log10(pval_mx))) + geom_point()
```

Write code to produce a Quantile-Quantile (QQ) plot for your p-values PLEASE NOTE (!!): do NOT use an R function (=write your own code to produce the Manhattan plot) but DO provide your code AND your QQ plot.

```{r}
heights <- phenotypes$Height

#when forming the quantiles we need two extra values for the -Inf and Inf on either end of the quantile range
normalQuantiles <- qnorm(seq(0,1,length.out = length(heights)+2),
                         mean = mean(heights), sd = sd(heights)) 

#then we remove the -Inf and Inf values
normalQuantiles <- normalQuantiles[2:(length(heights)+1)]

#the values have to be sorted so that each measured heights finds its corresponding theoretical quantile
qqDf <- data.frame(heights = sort(heights),
                   normalQuantiles = sort(normalQuantiles))
ggplot(qqDf, aes(heights, normalQuantiles)) + 
  geom_point() + 
  geom_abline(intercept = 0, slope = 1, color="red")
```
Do you consider the QQ plot to indicate that you have ‘good’ model fit in this case (and can therefore interpret the results of your analysis)? Explain your reasoning using no more than two sentences.

```
Yes, because most of the points roughly form a straight line. It has a 'good' model fit.
```

Write code that uses a Bonferroni correction to produce an overall study controlled Type I error of 0.05 to assess whether to reject the null hypothesis for each genotype, where your code also outputs the number of each genotype for which you rejected the null (remember: the genotypes are provided in order along the genome!). Report the numbers of all genotypes for which you rejected the null.

```{r}
adj_vals <- p.adjust(pval_mx, method = "bonferroni")
paste0("Significant p-values:", sum(adj_vals<0.05))

paste0("Genotype numbers for which I'm rejecting the null:", which(adj_vals<0.05))
```

Assuming the set of genotypes for which you rejected the null hypothesis in part [i] do indeed indicate the positions of causal genotypes in the genome, how many causal genotypes do you think these significant genotypes are indicating overall? Explain your reasoning using no more than two sentences.

```
We can reject the null hypothesis for one genotype. A genotype corresponds to a particular position on the genome and hence there can only be one causal genotype.
```
## 3 (Difficult)